presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
         }
     }
 
+    public static Map<List<String>, ComputedStatistics> groupComputedStatisticsByPartition(
+            List<ComputedStatistics> computedStatistics, List<String> partitionColumns, Map<String, Type> columnTypes)
+    {
+        List<Type> partitionColumnTypes = partitionColumns.stream()
+                .map(columnTypes::get)
+                .collect(toImmutableList());
+
+        return computedStatistics.stream()
+                .collect(toImmutableMap(statistics -> getPartitionValues(statistics, partitionColumns, partitionColumnTypes), statistics -> statistics));
+    }
+
+    private static List<String> getPartitionValues(ComputedStatistics statistics, List<String> partitionColumns, List<Type> partitionColumnTypes)
+    {
+        checkArgument(statistics.getGroupingColumns().equals(partitionColumns),
+                "Unexpected groping. Partition columns: %s. Grouping columns: %s", partitionColumns, statistics.getGroupingColumns());
+        Page partitionColumnsPage = new Page(1, statistics.getGropingValues().toArray(new Block[] {}));
+        return createPartitionValues(partitionColumnTypes, partitionColumnsPage, 0);
+    }
+
+    public static Map<String, HiveColumnStatistics> fromComputedStatistics(
+            ConnectorSession session, DateTimeZone timeZone, Map<ColumnStatisticMetadata, Block> computedStatistics, Map<String, Type> columnTypes, long rowCount)
+    {
+        return groupByColumn(computedStatistics)
+                .entrySet()
+                .stream()
+                .collect(toImmutableMap(Entry::getKey, entry -> createHiveColumnStatistics(session, timeZone, entry.getValue(), columnTypes.get(entry.getKey()), rowCount)));
+    }
+
+    private static Map<String, Map<ColumnStatisticType, Block>> groupByColumn(Map<ColumnStatisticMetadata, Block> computedStatistics)
+    {
+        Map<String, Map<ColumnStatisticType, Block>> result = new HashMap<>();
+        computedStatistics.forEach((metadata, block) -> {
```

Comment:
You might be able to use `Collectors.groupingBy` here

Reply:
I tried, but it really looks like an obfuscation then.